File: GoLang/src/Pieces/Piece.go
package Pieces

import (
	"fmt"
	"strings"
	"unicode"
)

func isUpper(s string) bool {
	for _, r := range s {
		if !unicode.IsUpper(r) && unicode.IsLetter(r) {
			return false
		}
	}
	return true
}

func inBounds(x int, y int) bool {
	return ((0 <= x) && (x < 8) && (0 <= y) && (y < 8))
}

func checkDuplicate(allowedMovements [][]int, x int, y int) bool {
	for index := 0; index < len(allowedMovements); index = index + 1 {
		if len(allowedMovements[index]) != 0 {
			if allowedMovements[index][0] != x && allowedMovements[index][1] != y {
				return true
			}
		}
	}
	return false
}

func walkable(side bool, board [8][8]string, x int, y int) bool {
	if inBounds(x, y) {
		toField := board[x][y]
		toFieldSide := isUpper(toField)
		if toField == "-" {
			return true
		}
		if toFieldSide == side {
			return false
		}
	}
	return true
}

func checkKeyList(side bool, board [8][8]string, move []int, allowedKeys [][]int) [][]int {
	// get baseline
	baseline := 1
	if !side {
		baseline = 6
	}
	// init some variables
	allowedMovements := make([][]int, cap(allowedKeys)*2+2)
	counter := 0
	var x, y, x1, x2, y1, y2 int
	// iterate the allowed keys
	for _, allowedKey := range allowedKeys {
		// calc the x and y coords for two steps and key steps
		x = move[1] - allowedKey[0]
		y = move[0] - allowedKey[1]
		if side {
			x1, y1 = x+3, y-1
			x2 = move[1] + allowedKey[0]
			y2 = move[0] + allowedKey[1]
		} else {
			x1, y1 = x-1, y-1
			x2 = move[1] - allowedKey[0]
			y2 = move[0] - allowedKey[1]
		}
		// add baseline 2 steps movement
		if baseline == move[1] && checkDuplicate(allowedMovements, x1, y1) {
			if walkable(side, board, x1, y1) {
				allowedMovements[counter] = []int{x1, y1}
			}
			counter += 1
		}
		// add keylist steps
		if walkable(side, board, x2, y2) {
			allowedMovements[counter] = []int{x2, y2}
		}
		// helpfull stuff
		counter += 1
	}
	//printTmpBoard(allowedMovements, board)
	return allowedMovements
}

func printTmpBoard(allowedMovements [][]int, board [8][8]string) {
	for index := 0; index < len(allowedMovements); index = index + 1 {
		if len(allowedMovements[index]) == 2 {
			x := allowedMovements[index][0]
			y := allowedMovements[index][1]
			if inBounds(x, y) {
				board[x][y] = "-"
			}
		}
	}
	print(allowedMovements)
	var i, j = 0, 0
	for ; i < 8; i++ {
		for j = 0; j < 8; j++ {
			fmt.Print(board[i][j] + "  ")
		}
		fmt.Println("")
	}
}

func checkIfAllowed(move []int, allowedMovements [][]int) bool {
	for index := 0; index < len(allowedMovements); index = index + 1 {
		if len(allowedMovements[index]) == 2 {
			x := allowedMovements[index][0]
			y := allowedMovements[index][1]
			if move[3] == x && move[2] == y {
				return true
			}
		}
	}
	return false
}

func pawnMovements(side bool, board [8][8]string, move []int) bool {
	allowedKeys := [][]int{{1, 0}, {1, 1}, {1, -1}}
	allowedMovements := checkKeyList(side, board, move, allowedKeys)
	return checkIfAllowed(move, allowedMovements)
}

func knightMovements(side bool, board [8][8]string, move []int) bool {
	allowedKeys := [][]int{{2, 1}, {2, -1}, {-2, -1}, {-2, 1},
		{1, 2}, {1, -2}, {-1, -2}, {-1, 2}}
	allowedMovements := checkKeyList(side, board, move, allowedKeys)
	return checkIfAllowed(move, allowedMovements)
}

func kingMovements(side bool, board [8][8]string, move []int) bool {
	allowedKeys := [][]int{{1, 1}, {1, -1}, {-1, 1}, {-1, -1},
		{1, 0}, {-1, 0}, {0, -1}, {0, 1}}
	allowedMovements := checkKeyList(side, board, move, allowedKeys)
	return checkIfAllowed(move, allowedMovements)
}

func ba(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x - 1
		y = y - 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func bb(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x + 1
		y = y + 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func bc(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x - 1
		y = y + 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func bd(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x + 1
		y = y - 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func bishopMovements(side bool, board [8][8]string, move []int) bool {
	allowedMovements := make([][]int, 32)
	counter := 0

	allowedMovements, counter = ba(move, counter, allowedMovements, board)
	allowedMovements, counter = bb(move, counter, allowedMovements, board)
	allowedMovements, counter = bc(move, counter, allowedMovements, board)
	allowedMovements, counter = bd(move, counter, allowedMovements, board)

	return checkIfAllowed(move, allowedMovements)
}

func ra(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x - 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func rb(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		x = x + 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func rc(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		y = y + 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func rd(move []int, counter int, allowedMovements [][]int,
	board [8][8]string) ([][]int, int) {
	x := move[1]
	y := move[0]
	for inBounds(x, y) {
		y = y - 1
		if inBounds(x, y) {
			if board[x][y] == "o" {
				allowedMovements[counter] = []int{x, y}
				counter += 1
				continue
			}
			if isUpper(board[x][y]) != isUpper(board[move[1]][move[0]]) {
				allowedMovements[counter] = []int{x, y}
				counter += 1
			} else {
				fmt.Println(x, y, board[x][y], board[move[1]][move[0]], "equal side")
				break
			}
		}
	}
	return allowedMovements, counter
}

func isFree(x int, y int, board [8][8]string, move []int, fromPiece string) bool {
	if board[x][y] == "o" {
		return true
	}
	if isUpper(fromPiece) != isUpper(board[x][y]) {
		return false
	}
	if isUpper(fromPiece) == isUpper(board[x][y]) {
		return true
	}
	return false
}

func rookMovements(side bool, board [8][8]string, move []int) bool {
	allowedMovements := make([][]int, 32)
	counter := 0

	allowedMovements, counter = ra(move, counter, allowedMovements, board)
	allowedMovements, counter = rb(move, counter, allowedMovements, board)
	allowedMovements, counter = rc(move, counter, allowedMovements, board)
	allowedMovements, counter = rd(move, counter, allowedMovements, board)

	return checkIfAllowed(move, allowedMovements)
}

func queenMovements(side bool, board [8][8]string, move []int) bool {
	allowedMovements := make([][]int, 32)
	counter := 0

	allowedMovements, counter = ba(move, counter, allowedMovements, board)
	allowedMovements, counter = bb(move, counter, allowedMovements, board)
	allowedMovements, counter = bc(move, counter, allowedMovements, board)
	allowedMovements, counter = bd(move, counter, allowedMovements, board)
	allowedMovements, counter = ra(move, counter, allowedMovements, board)
	allowedMovements, counter = rb(move, counter, allowedMovements, board)
	allowedMovements, counter = rc(move, counter, allowedMovements, board)
	allowedMovements, counter = rd(move, counter, allowedMovements, board)

	return checkIfAllowed(move, allowedMovements)
}

func CheckMovements(board [8][8]string, move []int) bool {
	fromPiece := board[move[1]][move[0]]
	result := false

	if fromPiece == "o" {
		return false
	}

	switch strings.ToLower(fromPiece) {
	case "p":
		result = pawnMovements(isUpper(fromPiece), board, move)
	case "k":
		result = knightMovements(isUpper(fromPiece), board, move)
	case "b":
		result = bishopMovements(isUpper(fromPiece), board, move)
	case "r":
		result = rookMovements(isUpper(fromPiece), board, move)
	case "q":
		result = queenMovements(isUpper(fromPiece), board, move)
	case "x":
		result = kingMovements(isUpper(fromPiece), board, move)
	}

	return result
}

File: Ruby/src/Pieces/Piece.rb
def check(movement, board, side) 
    fromPiece = board[movement[0]][movement[1]] 
    result = false
    if fromPiece == -6 || fromPiece == 6 
        result = check_pawn(movement, board, side)
    elsif fromPiece == -2 || fromPiece == 2 
        result = check_knight(movement, board, side)
    elsif fromPiece == -3 || fromPiece == 3 
        result = check_bishop(movement, board, side)
    elsif fromPiece == -1 || fromPiece == 1 
        result = check_rook(movement, board, side)
    elsif fromPiece == -4 || fromPiece == 4 
        result = check_queen(movement, board, side)
    elsif fromPiece == -5 || fromPiece == 5 
        result = check_king(movement, board, side)
    end
    return result
end

def key_list_check_pawn(allowed_movements, movement, board, side)
    for i in 0 ... allowed_movements.length
        x = movement[0] - allowed_movements[i][0]
        y = movement[1] - allowed_movements[i][1]
        if !side 
            x = movement[0] + allowed_movements[i][0]
            y = movement[1] + allowed_movements[i][1]
        end
        if 0 <= x and x < 8 and 0 <= y and y < 8
          if board[x][y] != 0 and (allowed_movements[i][1] != 0)
            return true 
          elsif board[x][y] == 0 and (allowed_movements[i][1] == 0)
            return true 
          end
        end
    end
    return false
end

def key_list_check(allowed_movements, movement, board, side)
    puts allowed_movements.length
    for i in 0 ... allowed_movements.length
        print allowed_movements[i], "\n"
        x = movement[0] - allowed_movements[i][0]
        y = movement[1] - allowed_movements[i][1]
        if !side 
            x = movement[0] + allowed_movements[i][0]
            y = movement[1] + allowed_movements[i][1]
        end
        if 0 <= x and x < 8 and 0 <= y and y < 8
            if board[x][y] > 0 and side
                return true
            elsif board[x][y] < 0 and !side
                return true
            elsif board[x][y] == 0
                return true
            end
        end
    end
    return false
end

def check_pawn(movement, board, side) 
    allowed_movements = [[1, 1], [1, -1], [1, 0]]
    x = movement[0] - 2 
    y = movement[1]
    if !side
        x = movement[0] + 2 
        y = movement[1] 
    end
    if movement[2] == x and movement[3] == y 
        return true
    end
    return key_list_check_pawn(allowed_movements, movement, board, side)
end

def check_knight(movement, board, side) 
    allowed_movements = [[2, 1], [2, -1], [-2, -1], [-2, 1], 
                         [1, 2], [1, -2], [-1, -2], [-1, 2]]
    return key_list_check(allowed_movements, movement, board, side)
end

def check_king(movement, board, side) 
    allowed_movements = [[1, 1], [1, -1], [-1, 1], [-1, -1], 
                         [1, 0], [-1, 0], [0, -1], [0, 1]]
    return key_list_check(allowed_movements, movement, board, side)
end

def check_queen(movement, board, side) 
    ba = ba(movement, board, side) 
    bbx = bb(movement, board, side) 
    bc = bc(movement, board, side) 
    bd = bd(movement, board, side) 
    ra = ra(movement, board, side) 
    rb = rb(movement, board, side) 
    rc = rc(movement, board, side) 
    rd = rd(movement, board, side) 
    return (ra or rb or rc or rd or ba or bbx or bc or bd) 
end

def check_bishop(movement, board, side) 
    ba = ba(movement, board, side) 
    bbx = bb(movement, board, side) 
    bc = bc(movement, board, side) 
    bd = bd(movement, board, side) 
    return (ba or bbx or bc or bd) 
end

def check_rook(movement, board, side) 
    ra = ra(movement, board, side) 
    rb = rb(movement, board, side) 
    rc = rc(movement, board, side) 
    rd = rd(movement, board, side) 
    return (ra or rb or rc or rd)
end

def ba(movement, board, side) 
    x = movement[0] + 1
    y = movement[1] + 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x + 1 
    y = y + 1 
    return false
end

def bb(movement, board, side) 
    x = movement[0] + 1
    y = movement[1] - 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x + 1 
    y = y - 1 
    return false
end

def bc(movement, board, side) 
    x = movement[0] - 1
    y = movement[1] + 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x - 1 
    y = y + 1 
    return false
end

def bd(movement, board, side) 
    x = movement[0] - 1
    y = movement[1] - 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x - 1 
    y = y - 1 
    return false
end

def ra(movement, board, side) 
    x = movement[0] - 1
    y = movement[1]
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x - 1 
    return false
end

def rb(movement, board, side) 
    x = movement[0] + 1
    y = movement[1]
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    x = x + 1 
    return false
end

def rc(movement, board, side) 
    x = movement[0]
    y = movement[1] - 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    y = y - 1 
    return false
end

def rd(movement, board, side) 
    x = movement[0] 
    y = movement[1] + 1
    while 0 <= x and x < 8 and 0 <= y and y < 8 
        if side 
            if board[x][y] < 0  
              return false
            elsif board[x][y] < 0 != board[movement[0]][movement[1]] < 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        else
            if board[x][y] > 0  
              return false
            elsif board[x][y] > 0 != board[movement[0]][movement[1]] > 0
              return true
            elsif board[x][y] == 0 
              return true
            end
        end
    end
    y = y + 1 
    return false
end



